Add overview documentation (#90)
kinobi-so/kinobi: README.md

# Kinobi

[![npm][npm-image]][npm-url]
[![npm-downloads][npm-downloads-image]][npm-url]
[![ci][ci-image]][ci-url]

[npm-downloads-image]: https://img.shields.io/npm/dm/kinobi.svg?style=flat
[npm-image]: https://img.shields.io/npm/v/kinobi.svg?style=flat
[npm-url]: https://www.npmjs.com/package/kinobi
[ci-image]: https://img.shields.io/github/actions/workflow/status/kinobi-so/kinobi/main.yml?logo=GitHub
[ci-url]: https://github.com/kinobi-so/kinobi/actions/workflows/main.yml

> Kinobi is a tool that describes any Solana program in a powerful standardised format known as the Kinobi IDL. This IDL can then be used to create a variety of utility such as rendering client code for your programs in various languages/frameworks, generating CLIs, providing more information to explorers, etc.

Reviewer: supper nit-pick: I think we don't need the ", etc." at the end of the sentence since we have "such as" just before providing the examples.

> ![Kinobi header: A small double-sided mind-map with the Kinobi logo in the middle. On the left, we see the various ways to get a Kinobi IDL from your Solana programs such as "Anchor Program" and "Shank macros". On the right, we see the various utility tools that are offered for the IDL such as "Rendering client code" or "Rendering documentation".](https://github.com/kinobi-so/kinobi/assets/3642397/1c3682f3-1cec-4ad4-a7b9-9f917a7a52f6)

Reviewer 2: * Render CLIs tools => Render CLIs or Render CLI tooling?

-   **From Anchor IDLs**. If you are using [Anchor programs](https://github.com/coral-xyz/anchor) or [Shank macros](https://github.com/metaplex-foundation/shank), then you can get an Anchor IDL from them. You can then use the `@kinobi-so/nodes-from-anchor` package to convert that IDL into a Kinobi IDL as shown in the code snippet below. Note that the Anchor IDL might not offer all the information that Kinobi can hold and therefore, you may want to transform your Kinobi IDL to provide additional information. You can learn more about this in the next section.

    ```ts
    import { createFromRootNode } from 'kinobi';
    import { rootNodeFromAnchor } from '@kinobi-so/nodes-from-anchor';
    import anchorIdl from 'anchor-idl.json';

    const kinobi = createFromRootNode(rootNodeFromAnchor(anchorIdl));
    ```

-   **By hand**. If your Solana program cannot be updated to use Kinobi macros and you don’t have an Anchor IDL, you may design your Kinobi IDL by hand. We may provide tools such as a Kinobi Playground to help with that in the future.

## Transforming Kinobi

> Once you have your Kinobi IDL, you may use visitors to transform it further. This can be useful when the Kinobi IDL was obtained from another source that may not contain as much information. Here is an example using two provided visitors that adjust the accounts and instructions on the program.

Reviewer 2: ```suggestion
Once you have your Kinobi IDL, you may use visitors to transform it further. This can be useful when the Kinobi IDL was obtained from another source that may not contain as much information. Here is an example using two provided visitors that adjusts the accounts and instructions on the program.
```

> Once you have your Kinobi IDL, you may use visitors to transform it further. This can be useful when the Kinobi IDL was obtained from another source that may not contain as much information. Here is an example using two provided visitors that adjust the accounts and instructions on the program.

Reviewer 2: ```suggestion
Once you have your Kinobi IDL, you may use visitors to transform it. This can be useful when the Kinobi IDL was obtained from another source that may not contain some necessary information. Here is an example using two provided visitors that adjust the accounts and instructions on the program.
```

```ts

> import { updateAccoutsVisitor, updateInstructionsVisitor } from 'kinobi';
> 
> kinobi.update(updateAccoutsVisitor({ ... }));

Reviewer 2: ```suggestion
import { updateAccountsVisitor, updateInstructionsVisitor } from 'kinobi';

kinobi.update(updateAccountsVisitor({ ... }));
```

kinobi.update(updateInstructionsVisitor({ ... }));
```

## From Kinobi to utility

Now that you have the perfect Kinobi IDL for your Solana program, you can benefit from all the visitors and tools that provide utility such as rendering client code or registering your IDL on-chain so explorers can dynamically display relevant information for your program.

_Note that some features such as rendering CLIs are not yet available. However, because the Kinobi IDL is designed as a tree of nodes, these features are only a visitor away from being ready. Feel free to reach out if you’d like to contribute to this Kinobi ecosystem._

-   **Rendering client code**. Want people to start interacting with your Solana program? You can use special visitors that go through your Kinobi IDL and generate client code that you can then publish for your end-users. Currently, we have the following renderers available:

>     -   `@kinobi-so/renderers-js`: Renders a JavaScript client compatible with the new [web3.js](https://github.com/solana-labs/solana-web3.js) (in technological preview).

Reviewer 2: ```suggestion
    -   `@kinobi-so/renderers-js`: Renders a JavaScript client compatible with the soon-to-be-released 2.0 line of [`@solana/web3.js`](https://github.com/solana-labs/solana-web3.js).
```
